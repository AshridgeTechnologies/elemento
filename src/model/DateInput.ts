import Element from './Element'
import {PropertyType} from './Types'
import BaseInputElement, {BaseInputProperties} from './BaseInputElement'

export type Properties = BaseInputProperties<Date> & {
}

export default class DateInput extends BaseInputElement<Properties> implements Element {

    readonly kind = 'DateInput'
    get iconClass() { return 'insert_invitation_outlined' }
    get valueType(): PropertyType { return 'date' }
}
