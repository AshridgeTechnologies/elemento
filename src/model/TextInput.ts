import BaseElement from './BaseElement'
import Element from './Element'
import {ElementId, PropertyValue} from './Types'

export type Properties = {
    readonly initialValue?: PropertyValue,
    readonly maxLength?: PropertyValue,
    readonly multiline?: PropertyValue,
    readonly label?: PropertyValue
}
export default class TextInput extends BaseElement<Properties> implements Element {

    constructor(
        id: ElementId,
        name: string,
        properties: Properties) {
        super(id, name, properties)
    }

    static is(element: Element): element is TextInput {
        return element.constructor.name === this.name
    }

    static defaultValue = ''

    get initialValue() { return this.properties.initialValue }
    get maxLength() { return this.properties.maxLength }
    get multiline() { return this.properties.multiline }
    get label() { return this.properties.label }
}