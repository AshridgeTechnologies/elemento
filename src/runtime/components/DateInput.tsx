import React, {ChangeEvent, FocusEvent} from 'react'
import {TextField} from '@mui/material'
import {definedPropertiesOf} from '../../util/helpers'
import {PropVal, valueOfProps} from '../runtimeFunctions'
import InputComponentState from './InputComponentState'
import {useGetObjectState} from '../appData'
import {DateType, NumberType} from '../../shared/types'
import {pick} from 'ramda'
import {InputWithInfo} from './InputWithInfo'
import {DatePicker} from '@mui/x-date-pickers'

type Properties = {path: string, label?: PropVal<string>, }

export default function DateInput({path, ...props}: Properties) {
    const {label} = valueOfProps(props)

    const state = useGetObjectState<DateInputState>(path)
    const {value} = state
    const dataType = state.dataType
    const required = dataType?.required
    const error = state.errorsShown && !state.valid
    const helperText = state.errorsShown && state.errors ? (state.errors as string[]).join('.  ') : undefined

    const onChange = (newDate: Date) => {
        state._setValue(newDate)
    }
    const onBlur = (_event: FocusEvent) => state.ShowErrors(true)

    const slotProps = {
        textField: {
            id: path,
            size: 'small' as 'small',
            error,
            helperText,
            onBlur
        }
    }
    const formControl = React.createElement(DatePicker, {
        label,
        slotProps,
        value,
        format: 'dd MMM yyyy',
        // @ts-ignore
        onChange,
        minDate: dataType?.min,
        maxDate: dataType?.max,
    })

    return InputWithInfo({description: dataType?.description, required, formControl})
}

export class DateInputState extends InputComponentState<Date, DateType>  {
    defaultValue = null
}

DateInput.State = DateInputState
