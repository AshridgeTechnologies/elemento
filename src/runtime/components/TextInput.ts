import React, {ChangeEvent} from 'react'
import {TextField} from '@mui/material'
import {definedPropertiesOf} from '../../util/helpers'
import {updateState} from '../appData'
import {valueOfProps} from '../runtimeFunctions'

type Properties = {state: {value?: string, _path: string, _controlValue: string | null}, label?: string, maxLength?: number, width?: string | number, multiline?: boolean}

export default function TextInput({state, ...props}: Properties) {
    const {maxLength, label, multiline, width} = valueOfProps(props)
    const inputProps = maxLength !== undefined ? {inputProps: {maxLength}} : {}
    const widthProp = width !== undefined ? {width} : {}
    const sxProps = {sx: {...widthProp}}
    const optionalProps = definedPropertiesOf({label, multiline})
    const {_path: path} = state
    const value = state._controlValue ?? ''
    const onChange = (event: ChangeEvent) => {
        const controlValue = (event.target as any).value
        const updateValue = controlValue !== '' ? controlValue : null
        updateState(path, {value: updateValue })
    }

    return React.createElement(TextField, {
        id: path,
        type: 'text',
        variant: 'outlined',
        size: 'small',
        value,
        onChange,
        ...inputProps,
        ...sxProps,
        ...optionalProps
    })
}
