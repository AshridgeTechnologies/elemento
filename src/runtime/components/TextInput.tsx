import React, {ChangeEvent, FocusEvent} from 'react'
import {TextField} from '@mui/material'
import {definedPropertiesOf} from '../../util/helpers'
import {PropVal, valueOfProps} from '../runtimeFunctions'
import InputComponentState from './InputComponentState'
import {useGetObjectState} from '../appData'
import {TextType} from '../../shared/types'
import {pick} from 'ramda'
import {InputWithInfo} from './InputWithInfo'


type Properties = {path: string, label?: PropVal<string>, width?: PropVal<string | number>, multiline?: PropVal<boolean>, readOnly?: PropVal<boolean>}

export default function TextInput({path, ...props}: Properties) {
    const {label, multiline: multilineProp, width, readOnly} = valueOfProps(props)
    const widthProp = width !== undefined ? {width} : {}
    const sxProps = {sx: {...widthProp}}

    const state = useGetObjectState<TextInputState>(path)
    const {value} = state
    const dataType = state.dataType
    const required = dataType?.required
    const multiline = dataType?.format === 'multiline' || multilineProp
    const multilineProps = multiline ? {minRows: 2, maxRows: 10} : {}
    const optionalProps = definedPropertiesOf({label, multiline, ...multilineProps})

    const dataTypeProps = definedPropertiesOf(pick(['maxLength', 'minLength'], dataType ?? {}))
    const format = dataType?.format
    if (format && ['url', 'email'].includes(format)) {
        dataTypeProps.type = format
    }
    const inputPropsValues = {...dataTypeProps, ...definedPropertiesOf({readOnly}), }
    const inputProps = Object.keys(inputPropsValues).length > 0 ? {inputProps: inputPropsValues} : {}

    const error = state.errorsShown && !state.valid
    const helperText = state.errorsShown && state.errors ? (state.errors as string[]).join('.  ') : undefined

    const onChange = (event: ChangeEvent) => {
        const controlValue = (event.target as any).value
        const updateValue = controlValue !== '' ? controlValue : null
        state._setValue(updateValue)
    }
    const onBlur = (_event: FocusEvent) => state.ShowErrors(true)

    const formControl =  React.createElement(TextField, {
        id: path,
        type: 'text',
        variant: 'outlined',
        size: 'small',
        value,
        error,
        helperText,
        onChange,
        onBlur,
        ...inputProps,
        ...sxProps,
        ...optionalProps
    })
    return InputWithInfo({description: dataType?.description, required, width, formControl})
}

export class TextInputState extends InputComponentState<string, TextType> {
    defaultValue = ''
}

TextInput.State = TextInputState