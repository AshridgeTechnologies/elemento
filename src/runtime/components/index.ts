export {default as App} from './App'
export {default as Button} from './Button'
export {default as Icon} from './Icon'
export {default as Image} from './Image'
export {default as UserLogon} from './UserLogon'
export {default as Menu} from './Menu'
export {default as MenuItem} from './MenuItem'
export {default as NumberInput} from './NumberInput'
export {default as DateInput} from './DateInput'
export {default as TrueFalseInput} from './TrueFalseInput'
export {default as SpeechInput} from './SpeechInput'
export {default as Data} from './Data'
export {default as Calculation} from './Calculation'
export {default as Timer} from './Timer'
export {default as Collection} from './Collection'
export {default as Page} from './Page'
export {default as ComponentElement} from './ComponentElement'
export {default as Form} from './Form'
export {default as BaseFormState} from './FormState'
export {default as SelectInput} from './SelectInput'
export {default as TextElement} from './TextElement'
export {default as Layout} from './Layout'
export {default as Block} from './Block'
export {default as AppBar} from './AppBar'
export {default as ListElement} from './ListElement'
export {default as ListItem} from './ListItem'
export {default as ItemSet} from './ItemSet'
export {default as ItemSetItem} from './ItemSetItem'
export {default as TextInput} from './TextInput'
export {default as MemoryDataStore} from './MemoryDataStore'
export {default as FileDataStore} from './FileDataStore'
export {default as FileDataStoreImpl} from './FileDataStoreImpl'
export {default as IdbDataStoreImpl} from './IdbDataStoreImpl'
export {default as BrowserDataStore} from './BrowserDataStore'
export {default as FirestoreDataStore} from './FirestoreDataStore'
export {default as FirestoreDataStoreImpl} from './FirestoreDataStoreImpl'
export {default as ServerAppConnector} from './ServerAppConnector'
export {default as Adapter} from './Adapter'
