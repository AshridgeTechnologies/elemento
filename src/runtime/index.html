<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="initial-scale=1, width=device-width" />
  <title>Elemento App</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700&display=swap"/>
</head>
<body>
  <div id="main"></div>
  <script src="https://unpkg.com/react@17/umd/react.development.js" crossorigin></script>
  <script src="https://unpkg.com/react-dom@17/umd/react-dom.development.js" crossorigin></script>
  <script src="runApp.ts" type="module"></script>
</body>
</html>