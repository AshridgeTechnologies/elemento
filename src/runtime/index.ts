import * as components from './components'
import {globalFunctions} from './globalFunctions'
export {components, globalFunctions}

export {default as React} from 'react'
export {default as ReactDOM} from 'react-dom/client'
export {default as appFunctions} from './appFunctions'
export {useGetObjectState, useObjectState} from './appData'
export {asArray, codeGenerationError, parentPath} from './runtimeFunctions'
export {default as AppRunner} from './AppRunner'
export {run} from './run'
