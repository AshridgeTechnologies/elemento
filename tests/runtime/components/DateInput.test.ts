/**
 * @vitest-environment jsdom
 */

import {expect, test} from "vitest"
import {DateInput} from '../../../src/runtime/components/index'
import {htmlSnapshot, snapshot, testAppInterface, wait, wrappedTestElement} from '../../testutil/testHelpers'
import {actWait, testContainer} from '../../testutil/rtlHelpers'
import {TextInputState} from '../../../src/runtime/components/TextInput'
import {DateInputState} from '../../../src/runtime/components/DateInput'
import {DateType} from '../../../src/runtime/types'
import {act, render, waitFor} from '@testing-library/react'

const [dateInput, appStoreHook] = wrappedTestElement(DateInput, DateInputState)

const stateAt = (path: string) => appStoreHook.stateAt(path)

const dateType1 = new DateType('startDate', {min: new Date('2020-01-01'), max: new Date('2021-12-31')})

test('DateInput element produces output with properties supplied',
    htmlSnapshot(dateInput('app.page1.start', {value: new Date('2021-11-12')}, {label: 'Starting Date', styles: {color: 'red'}}))
)

test('DateInput element produces readonly output with properties supplied',
    htmlSnapshot(dateInput('app.page1.start', {value: new Date('2021-11-12')}, {label: 'Starting Date', readOnly: true}))
)

test('DateInput element produces output with DataType supplied',
    htmlSnapshot(dateInput('app.page1.start1', {value: new Date('2021-11-12'), dataType: dateType1}, {label: 'Starting Date'}))
)

test('DateInput element produces output with default values where properties omittedxx',
    htmlSnapshot(dateInput('app.page1.finish'))
)

test('DateInput element produces output with default values where properties omitted',
    htmlSnapshot(dateInput('app.page1.finish'))
)

test('DateInput element produces output with description info', () => {
    const dateType = new DateType('tt1', {description: 'A date input for entering dates'})
    const {container} = render(dateInput('app.page1.description', {value: new Date('2021-11-12'), dataType: dateType}, {label: 'Starting Date'}))
    expect(container.innerHTML).toMatchSnapshot()
})

test('DateInput shows value from the state supplied', () => {
    const {el} = testContainer(dateInput('app.page1.widget1', {value: new Date('2021-11-12')}))
    expect(el`app.page1.widget1`.value).toBe('12 Nov 2021')
})

test('DateInput element produces output with properties supplied as state objects', () => {
    const {el} = testContainer(dateInput('app.page1.widget2', {value: new Date('2021-11-12')},
        {label: new TextInputState({value: 'Item Date'})}))
    expect(el`label[for="app.page1.widget2"]`.innerHTML).toBe('Item Date')
})

test('DateInput shows empty value when state value is absent', () => {
    const {el} = testContainer(dateInput('app.page1.widget1', {}))
    expect(el`app.page1.widget1`.value).toBe('')
})

test('DateInput shows empty value when state value is set to undefined', () => {
    const {el} = testContainer(dateInput('app.page1.widget3', {value: undefined}))
    expect(el`app.page1.widget3`.value).toBe('')
})

test('DateInput shows initial value when state value  exists', () => {
    const {el} = testContainer(dateInput('app.page1.widget4', {value: new Date('2021-11-12')}))
    expect(el`app.page1.widget4`.value).toBe('12 Nov 2021')
})

test('DateInput shows empty value when state value is set to null and initial value exists', () => {
    const {el}  = testContainer(dateInput('app.page1.widget5', new DateInput.State({value: new Date('2021-11-12')})._withStateForTest({value: null})))
    expect(el`app.page1.widget5`.value).toBe('')
})

test('DateInput stores updated values in the app store section for its path', async () => {
    const {enter}  = testContainer(dateInput('app.page1.singleDate', {value: new Date('2021-11-12')}))
    await enter('singleDate', '04 Mar 2022')
    expect(stateAt('app.page1.singleDate').value).toStrictEqual(new Date('2022-03-04'))
})

test('DateInput stores null value in the app store when cleared', async () => {
    const {el, user} = testContainer(dateInput('app.page1.termDate', {value: new Date('2021-11-12')}))
    const inputEl = el`[role=spinbutton]`
    await user.clear(inputEl)
    expect(stateAt('app.page1.termDate').dataValue).toBe(null)
})

test('State class has correct properties', () => {
    const emptyState = new DateInput.State({})
    expect(emptyState.value).toBe(null)
    expect(emptyState.dataValue).toBe(null)
    expect(emptyState.defaultValue).toBe(null)

    const state = new DateInput.State({value: new Date('2021-11-12')})
    const appInterface = testAppInterface('testPath', state)
    expect(state.value).toStrictEqual(new Date('2021-11-12'))
    expect(state.defaultValue).toBe(null)

    state.Reset()
    const resetState = state._withStateForTest({value: undefined, errorsShown: false})
    expect(appInterface.updateVersion).toHaveBeenCalledWith({value: undefined, errorsShown: false})
    expect(resetState.value).toStrictEqual(new Date('2021-11-12'))
    expect(resetState.dataValue).toStrictEqual(new Date('2021-11-12'))

    const clearedState = state._withStateForTest({value: null})
    expect(clearedState.value).toBe(null)
    expect(clearedState.dataValue).toBe(null)
})

test('DateInput stores errors shown when blurred and ShowErrors or Reset called', async () => {
    const {el, user, domContainer} = testContainer(dateInput('app.page1.startDate', {value: new Date('2020-05-06'), dataType: dateType1}))

    // expect(stateAt('app.page1.startDate').errorsShown).toBe(false)
    const inputEl = el`app.page1.startDate`
    await act(() => waitFor(() => { user.click(inputEl) }))
    await actWait(10)
    expect(stateAt('app.page1.startDate').errorsShown).toBe(false)

    await user.click(domContainer.ownerDocument.body)
    await wait(10)
    expect(stateAt('app.page1.startDate').errorsShown).toBe(true)

    await act( () => stateAt('app.page1.startDate').ShowErrors(false) )
    expect(stateAt('app.page1.startDate').errorsShown).toBe(false)

    await act( () => stateAt('app.page1.startDate').ShowErrors(true) )
    expect(stateAt('app.page1.startDate').errorsShown).toBe(true)

    await act( () => stateAt('app.page1.startDate').Reset() )
    expect(stateAt('app.page1.startDate').errorsShown).toBe(false)
})

test('DateInput shows errors if errorShown is true', async () => {
    const inputId = 'app.page1.stopDate'
    const {el} = testContainer(dateInput(inputId, {value: new Date('2019-05-06'), dataType: dateType1}, {label: 'The Input'}))

    await actWait(10)

    await actWait( () => stateAt(inputId).ShowErrors(false) )
    expect(el`p.MuiFormHelperText-root.Mui-error`).toBe(null)
    expect(el`label`.className.includes('Mui-error')).toBe(false)

    await actWait( () => stateAt(inputId).ShowErrors(true) )
    expect(el`p.MuiFormHelperText-root.Mui-error`.innerHTML).toBe('Earliest 01 Jan 2020')
    expect(el`label`.className.includes('Mui-error')).toBe(true)
})


