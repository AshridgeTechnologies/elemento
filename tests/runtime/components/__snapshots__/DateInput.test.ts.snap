// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateInput element produces output with DataType supplied 1`] = `
<div
  className="MuiFormControl-root MuiTextField-root css-350ohs-MuiFormControl-root-MuiTextField-root"
  onClick={[Function]}
  onMouseUp={[Function]}
>
  <label
    className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
    data-shrink={true}
    htmlFor="app.page1.start1"
    id="app.page1.start1-label"
  >
    Starting Date
  </label>
  <div
    className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      autoComplete="off"
      autoFocus={false}
      className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd css-19qh8xo-MuiInputBase-input-MuiOutlinedInput-input"
      disabled={false}
      id="app.page1.start1"
      inputMode="text"
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onPaste={[Function]}
      placeholder="DD MMMM YYYY"
      readOnly={false}
      required={false}
      type="text"
      value="12 Nov 2021"
    />
    <div
      className="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeSmall css-1laqsz7-MuiInputAdornment-root"
    >
      <button
        aria-label="Choose date, selected date is 12 Nov 2021"
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="CalendarIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
          />
        </svg>
      </button>
    </div>
    <fieldset
      aria-hidden={true}
      className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        className="css-14lo706"
      >
        <span>
          Starting Date
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;

exports[`DateInput element produces output with default values where properties omitted 1`] = `
<div
  className="MuiFormControl-root MuiTextField-root css-350ohs-MuiFormControl-root-MuiTextField-root"
  onClick={[Function]}
  onMouseUp={[Function]}
>
  <div
    className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      autoComplete="off"
      autoFocus={false}
      className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd css-19qh8xo-MuiInputBase-input-MuiOutlinedInput-input"
      disabled={false}
      id="app.page1.finish"
      inputMode="text"
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onPaste={[Function]}
      placeholder="DD MMMM YYYY"
      readOnly={false}
      required={false}
      type="text"
      value=""
    />
    <div
      className="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeSmall css-1laqsz7-MuiInputAdornment-root"
    >
      <button
        aria-label="Choose date"
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="CalendarIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
          />
        </svg>
        <span
          className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <fieldset
      aria-hidden={true}
      className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        className="css-ihdtdm"
      >
        <span
          className="notranslate"
        >
          â€‹
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;

exports[`DateInput element produces output with description info 1`] = `"<div class="MuiFormControl-root MuiTextField-root css-350ohs-MuiFormControl-root-MuiTextField-root"><label class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root" data-shrink="true" for="app.page1.description" id="app.page1.description-label">Starting Date</label><div class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"><input aria-invalid="false" autocomplete="off" id="app.page1.description" placeholder="DD MMMM YYYY" type="text" inputmode="text" class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd css-19qh8xo-MuiInputBase-input-MuiOutlinedInput-input" value="12 Nov 2021"><div class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeSmall css-1vu0h4l-MuiInputAdornment-root"><button class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-ea2eqt-MuiButtonBase-root-MuiIconButton-root" tabindex="0" type="button" aria-label="A date input for entering dates" data-mui-internal-clone-element="true"><span class="material-icons notranslate MuiIcon-root MuiIcon-fontSizeMedium css-kh3qk0-MuiIcon-root" aria-hidden="true" aria-label="Help">info_outlined</span><span class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"></span></button></div><fieldset aria-hidden="true" class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"><legend class="css-14lo706"><span>Starting Date</span></legend></fieldset></div></div>"`;

exports[`DateInput element produces output with properties supplied 1`] = `
<div
  className="MuiFormControl-root MuiTextField-root css-350ohs-MuiFormControl-root-MuiTextField-root"
  onClick={[Function]}
  onMouseUp={[Function]}
>
  <label
    className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
    data-shrink={true}
    htmlFor="app.page1.start"
    id="app.page1.start-label"
  >
    Starting Date
  </label>
  <div
    className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd css-ugk1i8-MuiInputBase-root-MuiOutlinedInput-root"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      autoComplete="off"
      autoFocus={false}
      className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd css-19qh8xo-MuiInputBase-input-MuiOutlinedInput-input"
      disabled={false}
      id="app.page1.start"
      inputMode="text"
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onPaste={[Function]}
      placeholder="DD MMMM YYYY"
      readOnly={false}
      required={false}
      type="text"
      value="12 Nov 2021"
    />
    <fieldset
      aria-hidden={true}
      className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        className="css-14lo706"
      >
        <span>
          Starting Date
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;

exports[`DateInput element produces readonly output with properties supplied 1`] = `
<div
  className="MuiFormControl-root MuiTextField-root css-350ohs-MuiFormControl-root-MuiTextField-root"
  onClick={[Function]}
  onMouseUp={[Function]}
>
  <label
    className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
    data-shrink={true}
    htmlFor="app.page1.start"
    id="app.page1.start-label"
  >
    Starting Date
  </label>
  <div
    className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd Mui-readOnly css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      autoComplete="off"
      autoFocus={false}
      className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd Mui-readOnly css-19qh8xo-MuiInputBase-input-MuiOutlinedInput-input"
      disabled={false}
      id="app.page1.start"
      inputMode="text"
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onPaste={[Function]}
      placeholder="DD MMMM YYYY"
      readOnly={true}
      required={false}
      style={
        {
          "z-index": 1,
        }
      }
      type="text"
      value="12 Nov 2021"
    />
    <fieldset
      aria-hidden={true}
      className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        className="css-14lo706"
      >
        <span>
          Starting Date
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;
